File: android/app/src/main/java/uz/minibozor/data/remote/dto/Dtos.kt
package uz.minibozor.data.remote.dto

import androidx.compose.runtime.Immutable
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// --------------------------------------------------------------------- shared

@Immutable
@Serializable
data class PageDto<T>(
    val items: List<T>,
    val page: Int,
    @SerialName("page_size") val pageSize: Int,
    val total: Int,
    @SerialName("has_more") val hasMore: Boolean,
)

@Immutable
@Serializable
data class MessageDto(val ok: Boolean = true, val message: String = "")

// ----------------------------------------------------------------------- auth

@Immutable
@Serializable
data class PhoneRequest(val phone: String)

@Immutable
@Serializable
data class OtpRequestedDto(
    val phone: String,
    @SerialName("expires_in") val expiresIn: Int,
    @SerialName("resend_after") val resendAfter: Int,
    @SerialName("dev_code") val devCode: String? = null,
)

@Immutable
@Serializable
data class OtpVerifyRequest(val phone: String, val code: String)

@Immutable
@Serializable
data class TokenPairDto(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String,
    @SerialName("expires_in") val expiresIn: Int,
    @SerialName("is_new_user") val isNewUser: Boolean = false,
)

@Immutable
@Serializable
data class RefreshRequest(@SerialName("refresh_token") val refreshToken: String)

@Immutable
@Serializable
data class PinRequest(val pin: String)

@Immutable
@Serializable
data class PinChangeRequest(
    @SerialName("current_pin") val currentPin: String? = null,
    @SerialName("new_pin") val newPin: String,
)

// ----------------------------------------------------------------------- user

@Immutable
@Serializable
data class UserDto(
    val id: Int,
    val phone: String,
    @SerialName("full_name") val fullName: String,
    val email: String? = null,
    @SerialName("birth_date") val birthDate: String? = null,
    val gender: String? = null,
    @SerialName("avatar_url") val avatarUrl: String? = null,
    val language: String,
    @SerialName("has_pin") val hasPin: Boolean,
    @SerialName("biometrics_enabled") val biometricsEnabled: Boolean,
)

@Immutable
@Serializable
data class UserUpdateRequest(
    @SerialName("full_name") val fullName: String? = null,
    val email: String? = null,
    @SerialName("birth_date") val birthDate: String? = null,
    val gender: String? = null,
)

@Immutable
@Serializable
data class SettingsDto(
    val language: String,
    @SerialName("location_enabled") val locationEnabled: Boolean,
    @SerialName("night_mode") val nightMode: Boolean,
)

@Immutable
@Serializable
data class SettingsRequest(
    val language: String? = null,
    @SerialName("location_enabled") val locationEnabled: Boolean? = null,
    @SerialName("night_mode") val nightMode: Boolean? = null,
)

@Immutable
@Serializable
data class NotificationPrefsDto(
    @SerialName("order_status") val orderStatus: Boolean,
    val promotions: Boolean,
    @SerialName("price_drop") val priceDrop: Boolean,
    val push: Boolean,
    val sms: Boolean,
)

@Immutable
@Serializable
data class NotificationPrefsRequest(
    @SerialName("order_status") val orderStatus: Boolean? = null,
    val promotions: Boolean? = null,
    @SerialName("price_drop") val priceDrop: Boolean? = null,
    val push: Boolean? = null,
    val sms: Boolean? = null,
)

@Immutable
@Serializable
data class ProfileOverviewDto(
    val user: UserDto,
    @SerialName("orders_count") val ordersCount: Int,
    @SerialName("favorites_count") val favoritesCount: Int,
    @SerialName("reviews_count") val reviewsCount: Int,
    @SerialName("addresses_count") val addressesCount: Int,
    @SerialName("cards_count") val cardsCount: Int,
    @SerialName("unread_notifications") val unreadNotifications: Int,
)

// -------------------------------------------------------------------- catalog

@Immutable
@Serializable
data class CategoryDto(
    val id: Int,
    val slug: String,
    val name: String,
    val subtitle: String = "",
    val icon: String = "box",
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("product_count") val productCount: Int = 0,
    @SerialName("has_children") val hasChildren: Boolean = false,
)

@Immutable
@Serializable
data class BrandDto(
    val id: Int,
    val slug: String,
    val name: String,
    @SerialName("product_count") val productCount: Int = 0,
)

@Immutable
@Serializable
data class VariantDto(
    val id: Int,
    val kind: String,
    val label: String,
    val value: String,
    /**
     * The product photographed in this colour, when the shop supplied one.
     *
     * A colour is chosen by looking at the thing rather than at a hex circle,
     * so the picker draws this when it is there and falls back to [value] when
     * it is not. Always null on a size.
     */
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("in_stock") val inStock: Boolean,
)

@Immutable
@Serializable
data class SpecDto(val key: String, val value: String)

@Immutable
@Serializable
data class ProductCardDto(
    val id: Int,
    val title: String,
    val price: Int,
    @SerialName("old_price") val oldPrice: Int? = null,
    @SerialName("discount_percent") val discountPercent: Int? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    val rating: Double = 0.0,
    @SerialName("reviews_count") val reviewsCount: Int = 0,
    val badge: String? = null,
    @SerialName("in_stock") val inStock: Boolean = true,
    @SerialName("is_favorite") val isFavorite: Boolean = false,
    /** Whether tapping "Savatga" should open the picker sheet or add at once. */
    @SerialName("has_variants") val hasVariants: Boolean = false,
)

@Immutable
@Serializable
data class ProductDto(
    val id: Int,
    val sku: String,
    val title: String,
    val subtitle: String = "",
    val description: String = "",
    val price: Int,
    @SerialName("old_price") val oldPrice: Int? = null,
    @SerialName("discount_percent") val discountPercent: Int? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    val images: List<String> = emptyList(),
    val rating: Double = 0.0,
    @SerialName("reviews_count") val reviewsCount: Int = 0,
    val badge: String? = null,
    @SerialName("in_stock") val inStock: Boolean = true,
    @SerialName("is_favorite") val isFavorite: Boolean = false,
    val category: CategoryDto,
    val brand: BrandDto? = null,
    val variants: List<VariantDto> = emptyList(),
    val specs: List<SpecDto> = emptyList(),
    val seller: String = "",
    val warranty: String? = null,
    @SerialName("stock_left") val stockLeft: Int = 0,
    @SerialName("is_original") val isOriginal: Boolean = true,
    @SerialName("free_delivery") val freeDelivery: Boolean = true,
    @SerialName("next_day_delivery") val nextDayDelivery: Boolean = true,
    @SerialName("delivery_note") val deliveryNote: String = "",
    /** How many have been sold — printed beside the rating. */
    @SerialName("sold_count") val soldCount: Int = 0,
)

@Immutable
@Serializable
data class BannerDto(
    val id: Int,
    val kicker: String = "",
    val title: String,
    val subtitle: String = "",
    val cta: String = "",
    @SerialName("image_url") val imageUrl: String,
    @SerialName("gradient_from") val gradientFrom: String,
    @SerialName("gradient_to") val gradientTo: String,
    @SerialName("target_type") val targetType: String,
    @SerialName("target_value") val targetValue: String,
)

@Immutable
@Serializable
data class SectionDto(
    val key: String,
    val title: String,
    val subtitle: String = "",
    val layout: String,
    @SerialName("category_slug") val categorySlug: String? = null,
    val products: List<ProductCardDto> = emptyList(),
)

@Immutable
@Serializable
data class HomeDto(
    val city: String,
    val banners: List<BannerDto>,
    val categories: List<CategoryDto>,
    val sections: List<SectionDto>,
)

@Immutable
@Serializable
data class FilterFlagDto(
    val key: String,
    val label: String,
    val subtitle: String = "",
    val count: Int = 0,
)

@Immutable
@Serializable
data class FiltersDto(
    @SerialName("price_min") val priceMin: Int,
    @SerialName("price_max") val priceMax: Int,
    val brands: List<BrandDto>,
    val sizes: List<String>,
    val ratings: List<String>,
    val flags: List<FilterFlagDto>,
    val sorts: List<Map<String, String>>,
)

@Immutable
@Serializable
data class SuggestionDto(
    @SerialName("product_id") val productId: Int,
    val title: String,
    val price: Int,
    @SerialName("image_url") val imageUrl: String? = null,
)

@Immutable
@Serializable
data class SearchLandingDto(val recent: List<String>, val popular: List<String>)

// -------------------------------------------------------------------- reviews

@Immutable
@Serializable
data class RatingBucketDto(val stars: Int, val count: Int, val percent: Int)

@Immutable
@Serializable
data class ReviewSummaryDto(
    val rating: Double,
    val total: Int,
    val distribution: List<RatingBucketDto>,
    /** A few customer photographs for the strip beside the rating. */
    val photos: List<String> = emptyList(),
    @SerialName("photos_total") val photosTotal: Int = 0,
)

@Immutable
@Serializable
data class ReviewDto(
    val id: Int,
    @SerialName("author_name") val authorName: String,
    @SerialName("author_initials") val authorInitials: String,
    val rating: Int,
    val text: String = "",
    @SerialName("variant_label") val variantLabel: String = "",
    val tags: List<String> = emptyList(),
    val photos: List<String> = emptyList(),
    val likes: Int = 0,
    @SerialName("liked_by_me") val likedByMe: Boolean = false,
    val status: String,
    @SerialName("created_at") val createdAt: String,
    val product: ProductCardDto? = null,
)

@Immutable
@Serializable
data class ReviewCreateRequest(
    val rating: Int,
    val text: String = "",
    val tags: List<String> = emptyList(),
    val photos: List<String> = emptyList(),
    @SerialName("variant_label") val variantLabel: String = "",
    @SerialName("order_item_id") val orderItemId: Int? = null,
)

// ----------------------------------------------------------------------- cart

@Immutable
@Serializable
data class CartItemDto(
    val id: Int,
    @SerialName("product_id") val productId: Int,
    val title: String,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("variant_label") val variantLabel: String = "",
    @SerialName("unit_price") val unitPrice: Int,
    @SerialName("old_unit_price") val oldUnitPrice: Int? = null,
    val quantity: Int,
    val selected: Boolean,
    @SerialName("in_stock") val inStock: Boolean,
    @SerialName("line_total") val lineTotal: Int,
)

@Immutable
@Serializable
data class CartTotalsDto(
    @SerialName("items_count") val itemsCount: Int,
    val subtotal: Int,
    val discount: Int,
    @SerialName("delivery_fee") val deliveryFee: Int,
    val total: Int,
    @SerialName("free_delivery_threshold") val freeDeliveryThreshold: Int,
    @SerialName("promo_code") val promoCode: String? = null,
)

@Immutable
@Serializable
data class CartDto(val items: List<CartItemDto>, val totals: CartTotalsDto)

@Immutable
@Serializable
data class CartAddRequest(
    @SerialName("product_id") val productId: Int,
    @SerialName("variant_id") val variantId: Int? = null,
    @SerialName("color_variant_id") val colorVariantId: Int? = null,
    val quantity: Int = 1,
)

@Immutable
@Serializable
data class CartUpdateRequest(val quantity: Int? = null, val selected: Boolean? = null)

@Immutable
@Serializable
data class PromoRequest(val code: String)

// ------------------------------------------------------------------- delivery

@Immutable
@Serializable
data class AddressDto(
    val id: Int,
    val title: String,
    val icon: String = "pin",
    val badge: String? = null,
    val line: String,
    val city: String = "",
    val meta: String = "",
    val floor: String? = null,
    val apartment: String? = null,
    @SerialName("entrance_code") val entranceCode: String? = null,
    val comment: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @SerialName("is_default") val isDefault: Boolean = false,
)

@Immutable
@Serializable
data class AddressRequest(
    val title: String = "Uy",
    val icon: String = "pin",
    val badge: String? = null,
    val line: String,
    val city: String = "Toshkent",
    val floor: String? = null,
    val apartment: String? = null,
    @SerialName("entrance_code") val entranceCode: String? = null,
    val comment: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @SerialName("is_default") val isDefault: Boolean = false,
)

@Immutable
@Serializable
data class PickupPointDto(
    val id: Int,
    val name: String,
    val address: String = "",
    val hours: String = "",
    @SerialName("distance_km") val distanceKm: Double? = null,
)

@Immutable
@Serializable
data class SlotDto(
    val id: Int,
    val day: String,
    @SerialName("start_time") val startTime: String,
    @SerialName("end_time") val endTime: String,
    val label: String,
    val note: String = "",
    val price: Int = 0,
    val express: Boolean = false,
    val available: Boolean = true,
)

@Immutable
@Serializable
data class SlotDayDto(
    val day: String,
    @SerialName("weekday_label") val weekdayLabel: String,
    @SerialName("day_label") val dayLabel: String,
    @SerialName("month_label") val monthLabel: String,
    val slots: List<SlotDto>,
)

// -------------------------------------------------------------------- payment

@Immutable
@Serializable
data class CardDto(
    val id: Int,
    val brand: String,
    val last4: String,
    val holder: String = "",
    val expiry: String,
    val status: String,
    @SerialName("is_default") val isDefault: Boolean = false,
)

@Immutable
@Serializable
data class CardRequest(
    val brand: String,
    val last4: String,
    val holder: String = "",
    @SerialName("expiry_month") val expiryMonth: Int,
    @SerialName("expiry_year") val expiryYear: Int,
    @SerialName("processor_token") val processorToken: String,
    @SerialName("is_default") val isDefault: Boolean = false,
)

// --------------------------------------------------------------------- orders

@Immutable
@Serializable
data class OrderItemDto(
    val id: Int,
    @SerialName("product_id") val productId: Int? = null,
    val title: String,
    @SerialName("image_url") val imageUrl: String = "",
    @SerialName("variant_label") val variantLabel: String = "",
    @SerialName("unit_price") val unitPrice: Int,
    val quantity: Int,
    @SerialName("line_total") val lineTotal: Int,
    val reviewed: Boolean = false,
)

@Immutable
@Serializable
data class OrderEventDto(
    val status: String,
    val title: String,
    @SerialName("happened_at") val happenedAt: String? = null,
    val note: String = "",
    val done: Boolean = false,
)

@Immutable
@Serializable
data class OrderSummaryDto(
    val id: Int,
    val code: String,
    val status: String,
    @SerialName("status_label") val statusLabel: String,
    val total: Int,
    @SerialName("items_count") val itemsCount: Int,
    @SerialName("preview_images") val previewImages: List<String> = emptyList(),
    @SerialName("eta_label") val etaLabel: String = "",
    @SerialName("created_at") val createdAt: String,
    @SerialName("can_cancel") val canCancel: Boolean = false,
    @SerialName("can_track") val canTrack: Boolean = false,
)

@Immutable
@Serializable
data class OrderDto(
    val id: Int,
    val code: String,
    val status: String,
    @SerialName("status_label") val statusLabel: String,
    val total: Int,
    @SerialName("items_count") val itemsCount: Int,
    @SerialName("preview_images") val previewImages: List<String> = emptyList(),
    @SerialName("eta_label") val etaLabel: String = "",
    @SerialName("created_at") val createdAt: String,
    @SerialName("can_cancel") val canCancel: Boolean = false,
    @SerialName("can_track") val canTrack: Boolean = false,
    @SerialName("delivery_kind") val deliveryKind: String,
    @SerialName("address_line") val addressLine: String = "",
    @SerialName("address_meta") val addressMeta: String = "",
    @SerialName("delivery_day") val deliveryDay: String? = null,
    @SerialName("delivery_start") val deliveryStart: String? = null,
    @SerialName("delivery_end") val deliveryEnd: String? = null,
    @SerialName("payment_method") val paymentMethod: String,
    @SerialName("payment_label") val paymentLabel: String = "",
    val paid: Boolean = false,
    @SerialName("recipient_name") val recipientName: String = "",
    @SerialName("recipient_phone") val recipientPhone: String = "",
    val subtotal: Int = 0,
    @SerialName("delivery_fee") val deliveryFee: Int = 0,
    val discount: Int = 0,
    val items: List<OrderItemDto> = emptyList(),
    val events: List<OrderEventDto> = emptyList(),
)

@Immutable
@Serializable
data class CheckoutRequest(
    @SerialName("address_id") val addressId: Int? = null,
    @SerialName("pickup_point_id") val pickupPointId: Int? = null,
    @SerialName("slot_id") val slotId: Int? = null,
    @SerialName("payment_method") val paymentMethod: String = "card",
    @SerialName("payment_card_id") val paymentCardId: Int? = null,
    @SerialName("recipient_name") val recipientName: String = "",
    @SerialName("recipient_phone") val recipientPhone: String = "",
    @SerialName("promo_code") val promoCode: String? = null,
    val comment: String = "",
)

@Immutable
@Serializable
data class CheckoutPreviewDto(
    val items: List<CartItemDto>,
    val address: AddressDto? = null,
    @SerialName("pickup_point") val pickupPoint: PickupPointDto? = null,
    val slot: SlotDto? = null,
    val card: CardDto? = null,
    val totals: CartTotalsDto,
)

@Immutable
@Serializable
data class ReasonDto(
    val id: Int,
    val label: String,
    @SerialName("requires_comment") val requiresComment: Boolean = false,
)

@Immutable
@Serializable
data class CancelRequest(
    @SerialName("reason_id") val reasonId: Int? = null,
    val reason: String = "",
    val comment: String = "",
)

@Immutable
@Serializable
data class ReturnRequestBody(
    @SerialName("order_item_id") val orderItemId: Int? = null,
    @SerialName("reason_id") val reasonId: Int? = null,
    val reason: String = "",
    val comment: String = "",
    val photos: List<String> = emptyList(),
)

@Immutable
@Serializable
data class ReturnDto(
    val id: Int,
    @SerialName("order_code") val orderCode: String,
    val reason: String,
    val comment: String = "",
    val status: String,
    @SerialName("created_at") val createdAt: String,
)

// ----------------------------------------------------------------------- misc

@Immutable
@Serializable
data class NotificationDto(
    val id: Int,
    val kind: String,
    val icon: String,
    val title: String,
    val text: String = "",
    @SerialName("deep_link") val deepLink: String? = null,
    val read: Boolean = false,
    @SerialName("created_at") val createdAt: String,
)

@Immutable
@Serializable
data class NotificationGroupDto(val label: String, val items: List<NotificationDto>)

@Immutable
@Serializable
data class FaqDto(val id: Int, val question: String, val answer: String = "")

@Immutable
@Serializable
data class LegalDocDto(
    val slug: String,
    val icon: String,
    val title: String,
    val meta: String = "",
)

@Immutable
@Serializable
data class LegalDocFullDto(
    val slug: String,
    val icon: String,
    val title: String,
    val meta: String = "",
    val body: String = "",
)

File: ios/MiniBozor/Data/DTO/DTOs.swift
import Foundation

// The decoder uses `.convertFromSnakeCase`, so these stay plain camelCase and
// need no CodingKeys.

// MARK: - Shared

struct PageDTO<T: Decodable>: Decodable {
    let items: [T]
    let page: Int
    let pageSize: Int
    let total: Int
    let hasMore: Bool
}

struct MessageDTO: Decodable {
    let ok: Bool
    let message: String
}

// MARK: - Auth

struct PhoneRequest: Encodable {
    let phone: String
}

struct OtpRequestedDTO: Decodable {
    let phone: String
    let expiresIn: Int
    let resendAfter: Int
    /// Only populated by dev builds of the backend.
    let devCode: String?
}

struct OtpVerifyRequest: Encodable {
    let phone: String
    let code: String
}

struct TokenPairDTO: Decodable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
    let isNewUser: Bool?
}

struct RefreshRequest: Encodable {
    let refreshToken: String
}

struct PinRequest: Encodable {
    let pin: String
}

struct PinChangeRequest: Encodable {
    let currentPin: String?
    let newPin: String
}

// MARK: - User

struct UserDTO: Decodable, Identifiable, Equatable {
    let id: Int
    let phone: String
    let fullName: String
    let email: String?
    let birthDate: String?
    let gender: String?
    let avatarUrl: String?
    let language: String
    let hasPin: Bool
    let biometricsEnabled: Bool
}

struct UserUpdateRequest: Encodable {
    var fullName: String?
    var email: String?
    var birthDate: String?
    var gender: String?
}

struct SettingsDTO: Decodable, Equatable {
    let language: String
    let locationEnabled: Bool
    let nightMode: Bool
}

struct SettingsRequest: Encodable {
    var language: String?
    var locationEnabled: Bool?
    var nightMode: Bool?
}

struct NotificationPrefsDTO: Decodable, Equatable {
    let orderStatus: Bool
    let promotions: Bool
    let priceDrop: Bool
    let push: Bool
    let sms: Bool
}

struct NotificationPrefsRequest: Encodable {
    var orderStatus: Bool?
    var promotions: Bool?
    var priceDrop: Bool?
    var push: Bool?
    var sms: Bool?
}

struct ProfileOverviewDTO: Decodable {
    let user: UserDTO
    let ordersCount: Int
    let favoritesCount: Int
    let reviewsCount: Int
    let addressesCount: Int
    let cardsCount: Int
    let unreadNotifications: Int
}

// MARK: - Catalog

struct CategoryDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let slug: String
    let name: String
    let subtitle: String
    let icon: String
    let imageUrl: String?
    let productCount: Int
    let hasChildren: Bool
}

struct BrandDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let slug: String
    let name: String
    let productCount: Int
}

struct VariantDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let kind: String
    let label: String
    let value: String
    /// The product photographed in this colour, so the picker can show the
    /// thing rather than ask the customer to imagine what `#0E0F12` looks like
    /// on a shoe. None for sizes, and for a colour nobody photographed.
    ///
    /// Optional rather than defaulted: a synthesised `Decodable` only falls
    /// back for optionals, so this is also what keeps an older backend from
    /// failing the whole product page over one missing key.
    let imageUrl: String?
    let inStock: Bool
}

struct SpecDTO: Decodable, Hashable {
    let key: String
    let value: String
}

struct ProductCardDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let title: String
    let price: Int
    let oldPrice: Int?
    let discountPercent: Int?
    let imageUrl: String?
    let rating: Double
    let reviewsCount: Int
    let badge: String?
    let inStock: Bool
    var isFavorite: Bool
    /// How many are left, so a tile can say when there are few.
    var stockLeft: Int = 0
    /// Whether tapping "Savatga" should open the picker sheet or add at once.
    var hasVariants: Bool = false
}

struct ProductDTO: Decodable, Identifiable {
    let id: Int
    let sku: String
    let title: String
    let subtitle: String
    let description: String
    let price: Int
    let oldPrice: Int?
    let discountPercent: Int?
    let imageUrl: String?
    let images: [String]
    let rating: Double
    let reviewsCount: Int
    let badge: String?
    let inStock: Bool
    var isFavorite: Bool
    let category: CategoryDTO
    let brand: BrandDTO?
    let variants: [VariantDTO]
    let specs: [SpecDTO]
    let seller: String
    let warranty: String?
    let stockLeft: Int
    let isOriginal: Bool
    let freeDelivery: Bool
    let nextDayDelivery: Bool
    let deliveryNote: String
    /// How many have been sold. The page prints it beside the rating, where
    /// "2 010 ta buyurtma" is the strongest thing on the panel.
    let soldCount: Int?

    var sizes: [VariantDTO] { variants.filter { $0.kind == "size" } }
    var colors: [VariantDTO] { variants.filter { $0.kind == "color" } }
    var sold: Int { soldCount ?? 0 }
}

struct BannerDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let kicker: String
    let title: String
    let subtitle: String
    let cta: String
    let imageUrl: String
    let gradientFrom: String
    let gradientTo: String
    let targetType: String
    let targetValue: String
}

struct SectionDTO: Decodable, Identifiable {
    var id: String { key }
    let key: String
    let title: String
    let subtitle: String
    let layout: String
    let categorySlug: String?
    let products: [ProductCardDTO]
}

struct HomeDTO: Decodable {
    let city: String
    let banners: [BannerDTO]
    let categories: [CategoryDTO]
    let sections: [SectionDTO]
}

struct FilterFlagDTO: Decodable, Identifiable, Hashable {
    var id: String { key }
    let key: String
    let label: String
    let subtitle: String
    let count: Int
}

struct FiltersDTO: Decodable {
    let priceMin: Int
    let priceMax: Int
    let brands: [BrandDTO]
    let sizes: [String]
    let ratings: [String]
    let flags: [FilterFlagDTO]
    let sorts: [[String: String]]
}

struct SuggestionDTO: Decodable, Identifiable {
    var id: Int { productId }
    let productId: Int
    let title: String
    let price: Int
    let imageUrl: String?
}

struct SearchLandingDTO: Decodable {
    let recent: [String]
    let popular: [String]
}

// MARK: - Reviews

struct RatingBucketDTO: Decodable, Identifiable {
    var id: Int { stars }
    let stars: Int
    let count: Int
    let percent: Int
}

struct ReviewSummaryDTO: Decodable {
    let rating: Double
    let total: Int
    let distribution: [RatingBucketDTO]
    /// A handful of customer photographs for the strip beside the rating, and
    /// the full count so the last tile can say how many more there are.
    let photos: [String]?
    let photosTotal: Int?

    var photoStrip: [String] { photos ?? [] }
    var photoCount: Int { photosTotal ?? 0 }
}

struct ReviewDTO: Decodable, Identifiable {
    let id: Int
    let authorName: String
    let authorInitials: String
    let rating: Int
    let text: String
    let variantLabel: String
    let tags: [String]
    let photos: [String]
    let likes: Int
    let likedByMe: Bool
    let status: String
    let createdAt: String
    let product: ProductCardDTO?
}

struct ReviewCreateRequest: Encodable {
    let rating: Int
    let text: String
    let tags: [String]
    let photos: [String]
    let variantLabel: String
    let orderItemId: Int?
}

// MARK: - Cart

struct CartItemDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let productId: Int
    let title: String
    let imageUrl: String?
    let variantLabel: String
    let unitPrice: Int
    let oldUnitPrice: Int?
    let quantity: Int
    let selected: Bool
    let inStock: Bool
    let lineTotal: Int
}

struct CartTotalsDTO: Decodable, Hashable {
    let itemsCount: Int
    let subtotal: Int
    let discount: Int
    let deliveryFee: Int
    let total: Int
    let freeDeliveryThreshold: Int
    let promoCode: String?
}

struct CartDTO: Decodable {
    let items: [CartItemDTO]
    let totals: CartTotalsDTO
}

struct CartAddRequest: Encodable {
    let productId: Int
    /// The size. A cart line carries a size *and* a colour.
    let variantId: Int?
    let colorVariantId: Int?
    let quantity: Int
}

struct CartUpdateRequest: Encodable {
    var quantity: Int?
    var selected: Bool?
}

struct PromoRequest: Encodable {
    let code: String
}

// MARK: - Delivery

struct AddressDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let title: String
    let icon: String
    let badge: String?
    let line: String
    let city: String
    let meta: String
    let floor: String?
    let apartment: String?
    let entranceCode: String?
    let comment: String?
    let latitude: Double?
    let longitude: Double?
    let isDefault: Bool
}

struct AddressRequest: Encodable {
    var title: String = L("preset_uy")
    var icon: String = "pin"
    var badge: String?
    var line: String
    var city: String = L("region_toshkent")
    var floor: String?
    var apartment: String?
    var entranceCode: String?
    var comment: String?
    var latitude: Double?
    var longitude: Double?
    var isDefault: Bool = false
}

struct PickupPointDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let name: String
    let address: String
    let hours: String
    let distanceKm: Double?
}

struct SlotDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let day: String
    let startTime: String
    let endTime: String
    let label: String
    let note: String
    let price: Int
    let express: Bool
    let available: Bool
}

struct SlotDayDTO: Decodable, Identifiable, Hashable {
    var id: String { day }
    let day: String
    let weekdayLabel: String
    let dayLabel: String
    let monthLabel: String
    let slots: [SlotDTO]
}

// MARK: - Payment

struct CardDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let brand: String
    let last4: String
    let holder: String
    let expiry: String
    let status: String
    let isDefault: Bool

    var isExpired: Bool { status == "expired" }
}

struct CardRequest: Encodable {
    let brand: String
    let last4: String
    let holder: String
    let expiryMonth: Int
    let expiryYear: Int
    /// The app never sends a card number — only the processor's token.
    let processorToken: String
    let isDefault: Bool
}

// MARK: - Orders

struct OrderItemDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let productId: Int?
    let title: String
    let imageUrl: String
    let variantLabel: String
    let unitPrice: Int
    let quantity: Int
    let lineTotal: Int
    let reviewed: Bool
}

struct OrderEventDTO: Decodable, Hashable, Identifiable {
    var id: String { status + title }
    let status: String
    let title: String
    let happenedAt: String?
    let note: String
    let done: Bool
}

struct OrderSummaryDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let code: String
    let status: String
    let statusLabel: String
    let total: Int
    let itemsCount: Int
    let previewImages: [String]
    let etaLabel: String
    let createdAt: String
    let canCancel: Bool
    let canTrack: Bool
}

struct OrderDTO: Decodable, Identifiable {
    let id: Int
    let code: String
    let status: String
    let statusLabel: String
    let total: Int
    let itemsCount: Int
    let previewImages: [String]
    let etaLabel: String
    let createdAt: String
    let canCancel: Bool
    let canTrack: Bool
    let deliveryKind: String
    let addressLine: String
    let addressMeta: String
    let deliveryDay: String?
    let deliveryStart: String?
    let deliveryEnd: String?
    let paymentMethod: String
    let paymentLabel: String
    let paid: Bool
    let recipientName: String
    let recipientPhone: String
    let subtotal: Int
    let deliveryFee: Int
    let discount: Int
    let items: [OrderItemDTO]
    let events: [OrderEventDTO]
}

struct CheckoutRequest: Encodable {
    var addressId: Int?
    var pickupPointId: Int?
    var slotId: Int?
    var paymentMethod: String = "card"
    var paymentCardId: Int?
    var recipientName: String = ""
    var recipientPhone: String = ""
    var promoCode: String?
    var comment: String = ""
}

struct CheckoutPreviewDTO: Decodable {
    let items: [CartItemDTO]
    let address: AddressDTO?
    let pickupPoint: PickupPointDTO?
    let slot: SlotDTO?
    let card: CardDTO?
    let totals: CartTotalsDTO
}

struct ReasonDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let label: String
    let requiresComment: Bool
}

struct CancelRequest: Encodable {
    var reasonId: Int?
    var reason: String = ""
    var comment: String = ""
}

struct ReturnRequestBody: Encodable {
    var orderItemId: Int?
    var reasonId: Int?
    var reason: String = ""
    var comment: String = ""
    var photos: [String] = []
}

struct ReturnDTO: Decodable, Identifiable {
    let id: Int
    let orderCode: String
    let reason: String
    let comment: String
    let status: String
    let createdAt: String
}

// MARK: - Misc

struct NotificationDTO: Decodable, Identifiable, Hashable {
    let id: Int
    let kind: String
    let icon: String
    let title: String
    let text: String
    let deepLink: String?
    let read: Bool
    let createdAt: String
}

struct NotificationGroupDTO: Decodable, Identifiable {
    var id: String { label }
    let label: String
    let items: [NotificationDTO]
}

struct FaqDTO: Decodable, Identifiable {
    let id: Int
    let question: String
    let answer: String
}

struct LegalDocDTO: Decodable, Identifiable {
    var id: String { slug }
    let slug: String
    let icon: String
    let title: String
    let meta: String
}

struct LegalDocFullDTO: Decodable {
    let slug: String
    let icon: String
    let title: String
    let meta: String
    let body: String
}
